README.md
# Bibata Cursor Rainbow

'Semi-Animated' Bibata cursors with rainbow colors.

[![build](https://github.com/ful1e5/Bibata_Cursor_Rainbow/actions/workflows/build.yml/badge.svg)](https://github.com/ful1e5/Bibata_Cursor_Rainbow/actions/workflows/build.yml)
[![CodeFactor](https://www.codefactor.io/repository/github/ful1e5/bibata_cursor_rainbow/badge)](https://www.codefactor.io/repository/github/ful1e5/bibata_cursor_rainbow)
[![Twitter](https://img.shields.io/badge/twitter-ful1e5-blue)](https://twitter.com/ful1e5)

### Bibata Styles

- **Bibata Original :** Sharp edge Bibata Cursors
- **Bibata Modern :** Round edge Bibata Cursors

### Cursor Sizes

<kbd>22</kbd>
<kbd>24</kbd>
<kbd>28</kbd>
<kbd>32</kbd>
<kbd>40</kbd>
<kbd>48</kbd>
<kbd>56</kbd>
<kbd>64</kbd>
<kbd>72</kbd>
<kbd>80</kbd>
<kbd>88</kbd>
<kbd>96</kbd>

### Quick install

<p align="center">
  <a href="https://www.pling.com/p/1445634/#files-panel" >
    <img title="Bibata Pling Store" width="40%" src="https://imgur.com/VxSgrWw.png">
  </a>
</p>

### Preview

<p align="center">
  <img title="Bibata Rainbow" width="90%" src="./svg/branding/preview.svg">
  </br>
  <sub>Bibata Rainbow</sub>
</p>

### Manual Install

Latest `Stable` & `Development` releases can be downloaded from [Here](https://github.com/ful1e5/Bibata_Cursor_Rainbow/releases)

#### Linux/X11

```bash
# extract `Bibata-Rainbow.tar.gz`
tar -xvf Bibata-Rainbow.tar.gz

# For local users
mv Bibata-Rainbow-* ~/.icons/

# For all users
sudo mv Bibata-Rainbow-* /usr/share/icons/
```

#### Windows

1. unzip `.zip` file
2. Open unziped directory in Explorer, and **right click** on `install.inf`.
3. Click 'Install' from the context menu, and authorize the modifications to your system.
4. Open _Control Panel_ > _Personalization and Appearance_ > _Change mouse pointers_, and select **Bibata Rainbow Cursors**.
5. Click '**Apply**'.

### Uninstall

#### Linux/X11

```bash
# For local users
rm ~/.icons/Bibata-Rainbow-*

# For all users
sudo rm /usr/share/icons/Bibata-Rainbow-*
```

#### Windows

1. Go to **Registry Editor** by typing the same in the _start search box_.
2. Expand `HKEY_CURRENT_USER` folder and expand `Control Panel` folder.
3. Go to `Cursors` folder and click on `Schemes` folder - all the available custom cursors that are installed will be listed here.
4. **Right Click** on the name of cursor file you want to uninstall; for eg.: _Bibata-Rainbow Cursors_ and click `Delete`.
5. Click '**yes**' when prompted.

# Dependencies

## External Libraries

- libxcursor-dev
- libx11-dev
- libpng-dev (<=1.6)

#### Install External Libraries

##### macOS

```bash
brew install --cask xquartz
brew install libpng
```

##### Debain/ubuntu

```bash
sudo apt install libx11-dev libxcursor-dev libpng-dev
```

##### ArchLinux/Manjaro

```bash
sudo pacman -S libx11 libxcursor libpng
```

##### Fedora/Fedora Silverblue/CentOS/RHEL

```bash
sudo dnf install libX11-devel libXcursor-devel libpng-devel
```

## Build Dependencies

- [gcc](https://gcc.gnu.org/install/)
- [make](https://www.gnu.org/software/make/)
- [nodejs](https://nodejs.org/en/) (<=12.x.x)
- [yarn](https://classic.yarnpkg.com/en/docs/install/)
- [python](https://www.python.org/downloads/) (<=3.8)
- [pip3](https://pip.pypa.io/en/stable/installing/)

### Node Packages

- [puppeteer](https://www.npmjs.com/package/puppeteer)
- [pngjs](https://www.npmjs.com/package/pngjs)
- [pixelmatch](https://www.npmjs.com/package/pixelmatch)

### PyPi Packages

- [clickgen](https://pypi.org/project/clickgen/s)

## Build From Scratch

### Auto Build (using GitHub Actions)

GitHub Actions is automatically runs on every `push`(on **main** & **dev** branch) and `pull request`(on **main** branch), You found theme resources in `artifact` section of **ci**. GitHub **Actions** available inside [.github/workflows](https://github.com/ful1e5/Bibata_Cursor_Rainbow/tree/main/.github/workflows) directory.

### Manual Build

> Check **[Makefile](./Makefile)** for more targets.

```bash
make
```

#### Build Only `XCursor` theme

```bash
make unix
```

#### Customize `XCursor` size

```bash
make unix X_SIZES=22            # Only built '22px' pixel-size.
make unix X_SIZES=22 24 32      # Multiple sizes are provided with  ' '(Space)
```

#### Install `XCursor` theme

```bash
make install            # install as user
  # OR
sudo make install       # install as root
```

#### Build Only `Windows` theme

```bash
make windows
```

#### Customize `Windows Cursor` size

```bash
make windows WIN_SIZE=96            # Supports only one pixel-size
```

> Windows installations steps are same as [these](#windows).

# You may also like...

- [**Bibata**](https://github.com/ful1e5/Bibata_Cursor) - Material based cursors
- [**Bibata Extra**](https://github.com/ful1e5/Bibata_Extra_Cursor) - More Bibata!
- [**Bibata Zebra**](https://github.com/ful1e5/Bibata-Zebra-Cursor) - Bibata cursor with a semi-animated strip
- [**Bibata Adapta**](https://gitlab.com/cscs/Bibata_AdaptaBreath_Cursors) - Bibata Based Cursor Made for AdaptaBreath and Manjaro.
- [**Bibata Translucent**](https://github.com/Silicasandwhich/Bibata_Cursor_Translucent) - Bibata translucent is a translucent flavor of the Bibata.

# Bugs

Bugs should be reported [here](https://github.com/ful1e5/Bibata_Cursor_Rainbow/issues) on the Github issues page.

# Getting Help

You can create a **issue**, I will help you.

# Contributing

Check [CONTRIBUTING.md](CONTRIBUTING.md), any suggestions for features and contributions to the continuing code masterelopment can be made via the issue tracker or code contributions via a `Fork` & `Pull requests`.

# Credit

- [Bibata](https://github.com/ful1e5/Bibata_Cursor)
- [Adwaita](https://github.com/GNOME/adwaita-icon-theme)
- [Dmz](https://github.com/GalliumOS/dmz-cursor-theme)
- [Yaru](https://github.com/ubuntu/yaru)
- Emojis are taken from [here](https://emojipedia.org/)
- Wedge loading from [loading.io](https://loading.io/spinner/wedges/-pie-wedge-pizza-circle-round-rotate) with **Microsoft** colors


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [unreleased]

### Added

- Bibata-Zebra link added
- pling link updated inside PLING.bbcode
- Add symlink for `top_left_arrow` cursor ful1e5/BreezeX_Cursor#10 ful1e5/BreezeX_Cursor#11
- Uninstall docs ful1e5/apple_cursor#79 ful1e5/apple_cursor#80

### Changed

- sponsorships updated
- minimal docs inside `README.md`
- Fix `None` value warning in `builder/symlinks.py`
- Github Action renamed to `build`

## [v1.1.2] - 25 Jul 2021

### Added

- Support button inside `PLING.bbcode` product page
- `make prepare` command for preparing bibata binaries
- `pyrightconfig.json` init

### Changed

- Removed **clean** target from `builder/Makefile`
- Compact code inside `builder/*`
- Remove `setup.py`
- Builder code moved to `src`
- Import `src` module directly inside `build.py`
- `Makefile` build commands re-arrange with groups
- Dynamic determine **Windows canvas size** on **Windows cursor size** inside `build.py`

## [v1.1.1] - 30 Mar 2021

### Added

- pling product page docs
- **2 space** format in `bitmapper`
- `brbpkg` installed as user in **make** command
- Supports clickgen v1.1.9
- `PNGProvider` imported from `clickgen.util`
- `builder/brbpkg` docstring in _reST_ with sphinx supports

### Changed

- python3 `virtualenv` commands removed from `builder/Makefile`
- Removed `builder/brbpkg/extra.py` module (all funtions _imported_ from `clickgen`)
- clean builder cache on every **make** commands

## [v1.1.0] - 04 Mar 2021

### Added

- Add Windows support
- `.svg` files separated to `<root>/svg`
- Customizable Windows & UNIX cursors
- Easy build system with `Make` supports
- Make docs in[README.md](./README.md#manual-build)
- use builtin clickgen WinPackager for Windows package
- Constant frames in animated cursors
- Smooth animation in Windows cursors
- `install` & `uninstall` with `make` commands

### Changed

- Individual cursor build with clickgen v1.1.8 #93
- puppeteer 'single-process' args removed
- All Bitmapping code structured in `bitmapper` directory
- builder code structured inside `builder` directory
- `builder` package renamed to `brbpkg`
- Symblinks provided by `brbpkg`(new Bibata Rainbow builder)
- Automatic build environment setup with `make`
- Logging file removed
- Fix dnf runtime requirements packages ful1e5/apple_cursor#34
- Fix grabbing cursors symlinks #87
- Fix Cursors not available for Login Window in Linux Mint 20 #88
- Dirty pixel in `move` cursor fixed #94

## [v1.0.0] - 17 Nov 2020

> First version of **Bibata_Cursor_Rainbow** is published on **Bibata_Cursor** repo

### Added

- **Bibata Rainbow** (semi-animated) cursors. **#55**
- **Bibata Rainbow** workflow file added `bibata-rainbow-ci.yml`
- **semi-animated** cursors `.svg` parsing support in `bibata-core`
- `PLING.bbcode` for Bibata Rainbow
- Tagging guid in [CHANGELOG.md](./CHANGELOG.md#tags)
- Bibata Rainbow's **branding** assets in `packages/rainbow/src/svgs/branding`

### Changed

- `build` workflow changed to `bibata-ci`
- **Bibata** workflow file names changed to `bibata-ci.yml` & `bibata-pr.yml`
- Optimize **BitmapsGenerator** browser.

[unreleased]: https://github.com/ful1e5/Bibata_Cursor_Rainbow/compare/v1.1.2...main
[v1.1.2]: https://github.com/ful1e5/Bibata_Cursor_Rainbow/compare/v1.1.2...v1.1.1
[v1.1.1]: https://github.com/ful1e5/Bibata_Cursor_Rainbow/compare/v1.1.1...v1.1.0
[v1.1.0]: https://github.com/ful1e5/Bibata_Cursor_Rainbow/tree/v1.1.0
[v1.0.0]: https://github.com/ful1e5/Bibata_Cursor/compare/v1.0.3...v1.0.0.br


builder/brbpkg/constants.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from typing import Dict

# Info
AUTHOR = "Kaiz Khatri"
URL = "https://github.com/ful1e5/Bibata_Cursor_Rainbow"

# XCursor
X_DELAY: int = 50


# Windows Cursor
WIN_DELAY = 2

X_CURSORS_CFG: Dict[str, Dict[str, int]] = {
    ##########
    # Static #
    ##########
    "crossed_circle.png": {"xhot": 100, "yhot": 100},
    "wayland-cursor.png": {"xhot": 100, "yhot": 100},
    "X_cursor.png": {"xhot": 100, "yhot": 100},
    ############
    # Animated #
    ############
    # Note: Animated cursors don't need an extension and frame numbers.
    "bd_double_arrow": {"xhot": 98, "yhot": 100},
    "bottom_left_corner": {"xhot": 31, "yhot": 172},
    "bottom_right_corner": {"xhot": 170, "yhot": 172},
    "bottom_side": {"xhot": 100, "yhot": 164},
    "bottom_tee": {"xhot": 100, "yhot": 164},
    "center_ptr": {"xhot": 98, "yhot": 131},
    "circle": {"xhot": 48, "yhot": 25},
    "context-menu": {"xhot": 48, "yhot": 25},
    "copy": {"xhot": 48, "yhot": 25},
    "cross": {"xhot": 98, "yhot": 96},
    # "crosshair": {"xhot": 99, "yhot": 99},
    "dnd_no_drop": {"xhot": 86, "yhot": 79},
    "dnd-ask": {"xhot": 86, "yhot": 79},
    "dnd-copy": {"xhot": 86, "yhot": 79},
    "dnd-link": {"xhot": 86, "yhot": 79},
    "dnd-move": {"xhot": 86, "yhot": 79},
    "dotbox": {"xhot": 100, "yhot": 100},
    "fd_double_arrow": {"xhot": 98, "yhot": 100},
    "grabbing": {"xhot": 106, "yhot": 79},
    "hand1": {"xhot": 113, "yhot": 95},
    "hand2": {"xhot": 88, "yhot": 32},
    "left_ptr": {"xhot": 53, "yhot": 36},
    "left_side": {"xhot": 35, "yhot": 100},
    "left_tee": {"xhot": 165, "yhot": 95},
    "link": {"xhot": 48, "yhot": 25},
    "ll_angle": {"xhot": 34, "yhot": 165},
    "lr_angle": {"xhot": 167, "yhot": 164},
    "move": {"xhot": 100, "yhot": 100},
    "pencil": {"xhot": 37, "yhot": 161},
    "plus": {"xhot": 100, "yhot": 100},
    "pointer-move": {"xhot": 48, "yhot": 25},
    "question_arrow": {"xhot": 102, "yhot": 102},
    "right_ptr": {"xhot": 150, "yhot": 29},
    "right_side": {"xhot": 163, "yhot": 98},
    "right_tee": {"xhot": 30, "yhot": 96},
    "sb_down_arrow": {"xhot": 100, "yhot": 126},
    "sb_h_double_arrow": {"xhot": 100, "yhot": 100},
    "sb_left_arrow": {"xhot": 86, "yhot": 100},
    "sb_right_arrow": {"xhot": 113, "yhot": 100},
    "sb_up_arrow": {"xhot": 99, "yhot": 86},
    "sb_v_double_arrow": {"xhot": 100, "yhot": 100},
    "tcross": {"xhot": 98, "yhot": 100},
    "top_left_corner": {"xhot": 29, "yhot": 27},
    "top_right_corner": {"xhot": 170, "yhot": 28},
    "top_side": {"xhot": 98, "yhot": 34},
    "top_tee": {"xhot": 98, "yhot": 29},
    "ul_angle": {"xhot": 34, "yhot": 35},
    "ur_angle": {"xhot": 164, "yhot": 34},
    "vertical-text": {"xhot": 100, "yhot": 100},
    "xterm": {"xhot": 100, "yhot": 100},
    "zoom-in": {"xhot": 90, "yhot": 89},
    "zoom-out": {"xhot": 93, "yhot": 90},
    # Truly animated
    "left_ptr_watch": {"xhot": 50, "yhot": 28, "delay": 25},
    "wait": {"xhot": 100, "yhot": 100, "delay": 25},
}

WIN_CURSORS_CFG: Dict[str, Dict[str, str]] = {
    "right_ptr": {"to": "Alternate", "position": "top_left"},
    "cross": {"to": "Cross"},
    "left_ptr": {"to": "Default", "position": "top_left"},
    "bd_double_arrow": {"to": "Diagonal_1"},
    "fd_double_arrow": {"to": "Diagonal_2"},
    "pencil": {"to": "Handwriting"},
    "question_arrow": {"to": "Help", "position": "top_left"},
    "sb_h_double_arrow": {"to": "Horizontal"},
    "xterm": {"to": "IBeam", "position": "top_left"},
    "hand2": {"to": "Link", "position": "top_left"},
    "hand1": {"to": "Move"},
    "circle": {"to": "Unavailiable", "position": "top_left"},
    "sb_v_double_arrow": {"to": "Vertical"},
    "left_ptr_watch": {"to": "Work", "position": "top_left"},
    "wait": {"to": "Busy"},
}
